R/type-list-of.R

```diff
@@ -188,21 +400,56 @@ vec_ptype2.vctrs_list_of.vctrs_list_of <- function(
   x_arg = "",
   y_arg = ""
 ) {
-  x_ptype <- attr(x, "ptype", exact = TRUE)
-  y_ptype <- attr(y, "ptype", exact = TRUE)
-  if (identical(x_ptype, y_ptype)) {
+  x_ptype <- list_of_ptype(x)
+  y_ptype <- list_of_ptype(y)
+
+  x_size <- list_of_size(x)
+  y_size <- list_of_size(y)
+
+  if (identical(x_ptype, y_ptype) && identical(x_size, y_size)) {
     return(x)
   }
 
-  tryCatch(
-    expr = {
-      ptype <- vec_ptype2(x_ptype, y_ptype, x_arg = x_arg, y_arg = y_arg)
-      new_list_of0(x = list(), ptype = ptype)
-    },
-    vctrs_error_incompatible_type = function(cnd) {
-      list()
-    }
-  )
+  # Common type always goes towards more lenient type
+  #
+  # Element type:
+  # - If either `x_ptype` or `y_ptype` are `NULL`, fall back to `NULL`
+  # - If both `x_ptype` and `y_ptype` are specified, try common type but fall
+  #   back to `NULL`
+  #
+  # Element size:
+  # - If either `x_size` or `y_size` are `NULL`, fall back to `NULL`
+  # - If both `x_ptype` and `y_ptype` are specified, try common size but fall
+  #   back to `NULL`
+  #
+  # If both `ptype` and `size` are `NULL` after this, return bare `list()`,
+  # otherwise return `list_of()` with appropriate restrictions. Note that with
+  # this set up we may fail a ptype2 determination but pass a size2
+  # determination and still return a list-of.
+  if (is_null(x_ptype) || is_null(y_ptype)) {
+    ptype <- NULL
+  } else {
+    ptype <- tryCatch(
+      vec_ptype2(x_ptype, y_ptype),
+      vctrs_error_incompatible_type = function(cnd) NULL
+    )
+  }
+
+  if (is_null(x_size) || is_null(y_size)) {
+    size <- NULL
+  } else {
+    # No `vec_size2()`. This uses ALTREP to be efficient.
+    size <- tryCatch(
+      vec_size_common(seq_len(x_size), seq_len(y_size)),
+      vctrs_error_incompatible_size = function(cnd) NULL
+    )
```

Comment:
This is admittedly a bit hacky...but it is fast? It never materializes the object.